### PR TITLE
ClientUserCodeDeployment feature development

### PR DESCRIPTION
**Pull request by sancar**

### CODE REVIEW

**Comment by sancar** [comment]
run-lab-run

File: hazelcast-client/src/main/java/com/hazelcast/client/impl/HazelcastClientInstanceImpl.java [hunk 1]
```diff
@@ -411,16 +414,17 @@ public void start() {
         metricsRegistry.collectMetrics(listenerService);
 
         try {
+            userCodeDeploymentService.start();
+            listenerService.start();
```
Comment (ihsandemir) [inline hazelcast-client/src/main/java/com/hazelcast/client/impl/HazelcastClientInstanceImpl.java]: why did you change start order between listenerService and clusterService?

File: hazelcast-client/src/main/java/com/hazelcast/client/spi/impl/ClientUserCodeDeploymentService.java [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+import java.util.jar.JarEntry;
+import java.util.jar.JarInputStream;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import static com.hazelcast.nio.IOUtil.toByteArray;
+
+public class ClientUserCodeDeploymentService {
+
+    private static final Pattern CLASS_PATTERN = Pattern.compile("(.*)\\.class$");
```
Comment (ihsandemir) [inline hazelcast-client/src/main/java/com/hazelcast/client/spi/impl/ClientUserCodeDeploymentService.java]: is the inner classes listed as a name in a jar file list? if so, do they match the pattern?

File: hazelcast-client/src/test/java/com/hazelcast/client/usercodedeployment/ClientUserCodeDeploymentTest.java [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+                .setEnabled(true)
+                .setClassCacheMode(classCacheMode)
+                .setProviderMode(providerMode);
+        return config;
+    }
+
+    private ClientConfig createClientConfig() {
+        ClientConfig config = new ClientConfig();
+        ClientUserCodeDeploymentConfig clientUserCodeDeploymentConfig = new ClientUserCodeDeploymentConfig();
+        clientUserCodeDeploymentConfig.addClass("usercodedeployment.IncrementingEntryProcessor");
```
Comment (ihsandemir) [inline hazelcast-client/src/test/java/com/hazelcast/client/usercodedeployment/ClientUserCodeDeploymentTest.java]: Can you also add tests for addJar API? I see them exist in the ClientUserCodeDeploymentConfigTest but using here also will test the end-to-end feature.

File: hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/DeployClassesMessageTask.java [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+    @Override
+    protected Supplier<Operation> createOperationSupplier() {
+        return this;
+    }
+
+    @Override
+    protected Object reduce(Map<Member, Object> map) throws Throwable {
+        for (Object result : map.values()) {
+            if (result instanceof Throwable && !(result instanceof MemberLeftException)) {
+                throw (Throwable) result;
```
Comment (ihsandemir) [inline hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/DeployClassesMessageTask.java]: What if a node produces IllegalStateException due to binary mismatch and another node which does not even have the class (getting it the first time), deploys it correctly? Should we revert the changes at the correctly deployed members?

Comment (sancar) [inline hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/DeployClassesMessageTask.java]: Yes, user needs to revert the changes acoording to current design. Members needs to restarted. If we had a undeploy in the future, it can also be used without restarting members.

File: hazelcast-client/src/test/java/com/hazelcast/client/usercodedeployment/ClientUserCodeDeploymentTest.java (hunk 3 above)
Comment (sancar) [inline hazelcast-client/src/test/java/com/hazelcast/client/usercodedeployment/ClientUserCodeDeploymentTest.java]: I am adding a test

File: hazelcast-client/src/main/java/com/hazelcast/client/spi/impl/ClientUserCodeDeploymentService.java (hunk 2 above)
Comment (sancar) [inline hazelcast-client/src/main/java/com/hazelcast/client/spi/impl/ClientUserCodeDeploymentService.java]: I am adding a test with inner and anonymous classes

File: hazelcast-client/src/test/java/com/hazelcast/client/usercodedeployment/ClientUserCodeDeploymentTest.java [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+         import java.util.Map;
+
+        public class IncrementingEntryProcessor extends AbstractEntryProcessor<Integer, Integer> {
+
+            @Override
+            public Object process(final Map.Entry<Integer, Integer> entry) {
+                Integer origValue = entry.getValue();
+                final Integer newValue = origValue + 1;
+
+                final usercodedeployment.IncrementingEntryProcessor.Test test = new usercodedeployment.IncrementingEntryProcessor.Test(entry);
```
Comment (sancar) [inline hazelcast-client/src/test/java/com/hazelcast/client/usercodedeployment/ClientUserCodeDeploymentTest.java]: I am getting following exception here
```
java.lang.IllegalAccessError: usercodedeployment/IncrementingEntryProcessor$Test
	at usercodedeployment.IncrementingEntryProcessor.process(IncrementingEntryProcessor.java:35) ~[test-classes/:?]
	at com.hazelcast.map.impl.operation.EntryOperator.process(EntryOperator.java:328) ~[classes/:?]
	at com.hazelcast.map.impl.operation.EntryOperator.operateOnKeyValueInternal(EntryOperator.java:187) ~[classes/:?]
	at com.hazelcast.map.impl.operation.EntryOperator.operateOnKey(EntryOperator.java:172) ~[classes/:?]
	at com.hazelcast.map.impl.operation.PartitionWideEntryOperation.run(PartitionWideEntryOperation.java:84) ~[classes/:?]
	at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:194) [classes/:?]
	at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.process(OperationThread.java:120) [classes/:?]
	at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.run(OperationThread.java:100) [classes/:?]
```
I verified that UserCodeDeploymentClassLoader is able to find the class definition of inner class Test and return.
But as soon as loadClass method returns it throw the exception above. 

@jerrinot any idea ?

**Comment by sancar** [comment]
run-lab-run

**Comment by sancar** [comment]
run-lab-run

**Comment by sancar** [comment]
run-lab-run

**Comment by sancar** [comment]
run-lab-run

**Comment by sancar** [comment]
run-lab-run

File: hazelcast-client/src/main/java/com/hazelcast/client/spi/impl/ClientUserCodeDeploymentService.java (hunk 2 above)
Comment (sancar) [inline hazelcast-client/src/main/java/com/hazelcast/client/spi/impl/ClientUserCodeDeploymentService.java]: Inner and anonymous classes failed because of another reason. I prepared the test but We will send the fix and test as seperate pr.

File: hazelcast-client/src/main/java/com/hazelcast/client/spi/impl/ClientUserCodeDeploymentService.java [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+            } while (true);
+        } finally {
+            IOUtil.closeResource(inputStream);
+        }
+    }
+
+    private JarInputStream getJarInputStream(String jarPath) throws IOException {
+        File file = new File(jarPath);
+        if (file.exists()) {
+            return new JarInputStream(new FileInputStream(jarPath));
```
Comment (mdogan) [inline hazelcast-client/src/main/java/com/hazelcast/client/spi/impl/ClientUserCodeDeploymentService.java]: `new FileInputStream(file)`

File: hazelcast-client/src/main/java/com/hazelcast/client/spi/impl/ClientUserCodeDeploymentService.java [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+        }
+
+        InputStream inputStream = configClassLoader.getResourceAsStream(jarPath);
+        if (inputStream == null) {
+            throw new FileNotFoundException("File could not be found in " + jarPath
```
Comment (mdogan) [inline hazelcast-client/src/main/java/com/hazelcast/client/spi/impl/ClientUserCodeDeploymentService.java]: What if `jarPath ` is a URL?

Comment (sancar) [inline hazelcast-client/src/main/java/com/hazelcast/client/spi/impl/ClientUserCodeDeploymentService.java]: 👍  I am adding URL handling to handle following usage
`addJar(url.toExternalForm())`

File: hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/DeployClassesMessageTask.java [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+public class DeployClassesMessageTask extends AbstractMultiTargetMessageTask<ClientDeployClassesCodec.RequestParameters>
+        implements Supplier<Operation> {
+
+    public DeployClassesMessageTask(ClientMessage clientMessage, Node node, Connection connection) {
+        super(clientMessage, node, connection);
+    }
+
+    @Override
+    public Operation get() {
+        return new DeployClassesOperation(parameters.classDefinitions);
```
Comment (jerrinot) [inline hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/DeployClassesMessageTask.java]: shouldn't this eagerly check cluster protocol version? when running in 3.8 mode then it should fail-fast. 
otherwise the operation can fail during deserialization on remote members.

Comment (sancar) [inline hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/DeployClassesMessageTask.java]: @jerrinot I have updated the createOperationSupplier method in DeployClassesMessageTask accordingly.
